### PR TITLE
feat: Support transient identities and traits

### LINKED ISSUE #4279: Support transient identities and traits in Core API
Part of #4278.

Evolve the API schemas as per the following:

### `GET /api/v1/identities` ([docs](https://api.flagsmith.com/api/v1/docs/#/api/identify_user)):

```diff
{
    "get": {
        "operationId": "identify_user",
        "description": "",
        "parameters": [
            {
                "name": "identifier",
                "in": "query",
                "required": true,
                "type": "string",
                "minLength": 1
+            },
+            {
+                "name": "transient",
+                "in": "query",
+                "required": false,
+                "type": "boolean"
            }
        ],
        "responses": {
            "200": {
                "description": "",
                "schema": {
                    "$ref": "#/definitions/SDKIdentitiesResponse"
                }
            }
        },
        "tags": [
            "api"
        ]
    }
```

#### `#/definitions/IdentifyWithTraits`

```diff
{
    "required": [
        "identifier"
    ],
    "type": "object",
    "properties": {
        "identifier": {
            "title": "Identifier",
            "type": "string",
            "minLength": 1
        },
        "traits": {
            "type": "array",
            "items": {
                "$ref": "#/definitions/TraitSerializerBasic"
            }
        },
        "flags": {
            "type": "array",
            "items": {
                "$ref": "#/definitions/SDKFeatureState"
            },
            "readOnly": true
+        },
+        "transient": {
+            "type": "boolean",
+            "x-nullable": true
        }
    }
}
```

#### `#/definitions/TraitSerializerBasic`

```diff
{
    "required": [
        "trait_key",
        "trait_value"
    ],
    "type": "object",
    "properties": {
        "id": {
            "title": "ID",
            "type": "integer",
            "readOnly": true
        },
        "trait_key": {
            "title": "Trait key",
            "type": "string",
            "maxLength": 200,
            "minLength": 1
        },
        "trait_value": {
            "title": "Trait value",
            "type": "string",
            "x-nullable": true
+        },
+        "transient": {
+            "type": "boolean",
+            "x-nullable": true
        }
    }
}
```

When endpoint gets called:
- If identity is marked as transient, return the evaluated flags, but don't persist it.
- If a trait is marked as transient, use it when evaluating flags, but don't persist it.

### PR DESCRIPTION
**Pull request by khvn26**
Thanks for submitting a PR! Please check the boxes below:

- [x] I have run [`pre-commit`](https://docs.flagsmith.com/platform/contributing#pre-commit) to check linting
- [x] I have added information to `docs/` if required so people know about the feature!
- [x] I have filled in the "Changes" section below?
- [x] I have filled in the "How did you test this code" section below?
- [x] I have used a [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) title for this Pull Request

## Changes

Closes #4279.
Contributes to #4278.

This adds support for
- Top-level `transient` parameter
- `transient` attribute for traits

for the `/api/v1/identities` SDK endpoint (GET, POST).

When including top-level `transient: true`, identity is evaluated against segments but is not persisted at all.
When calling the endpoint with traits marked as transient, these traits get evaluated against segments but do not get persisted.

Additionally, `Identity.update_traits` does not perform two `SELECT` queries anymore.

## How did you test this code?

Added unit tests for views and serializers.

### CODE REVIEW

File: api/environments/identities/models.py [hunk 1]
```diff
@@ -235,23 +240,30 @@ def update_traits(self, trait_data_items):
...
         for trait_data_item in trait_data_items:
             trait_key = trait_data_item["trait_key"]
             trait_value = trait_data_item["trait_value"]
+            transient = trait_data_item.get("transient")
 
             if trait_value is None:
                 # build a list of trait keys to delete having been nulled by the
                 # input data
-                keys_to_delete.append(trait_key)
+                keys_to_delete.add(trait_key)
```
Comment (matthewelwell) [inline api/environments/identities/models.py:267]: This is an interesting design question (but a fairly slim use case). If a trait is marked as transient, and set to null, should it be able to delete an existing persisted trait 🤔 ?

File: api/environments/identities/models.py [hunk 2]
```diff
@@ -278,5 +296,15 @@ def update_traits(self, trait_data_items):
         Trait.objects.bulk_create(new_traits, ignore_conflicts=True)
 
         # return the full list of traits for this identity by refreshing from the db
-        # TODO: handle this in the above logic to avoid a second hit to the DB
-        return self.identity_traits.all()
+        # override persisted traits by transient traits in case of key collisions
```
Comment (matthewelwell) [inline api/environments/identities/models.py:308]: Agreed 👍

File: api/environments/identities/models.py [hunk 3]
```diff
@@ -278,5 +296,15 @@ def update_traits(self, trait_data_items):
...
+        return [
+            *{
+                trait.trait_key: trait
+                for trait in chain(
+                    current_traits.values(),
+                    updated_traits,
+                    new_traits,
+                    transient_traits,
+                )
```
Comment (matthewelwell) [inline api/environments/identities/models.py:317]: Looks like this removes a query 👀 👌

File: api/tests/unit/environments/identities/test_unit_identities_models.py [hunk 4]
```diff
@@ -662,19 +662,33 @@ def test_update_traits(environment: Environment) -> None:
...
+    # and one transient trait that is evaluated against but not persisted
+    transient_trait_key = "transient"
+    transient_trait_value = "not persisted"
+
     trait_data_items = [
         generate_trait_data_item(
             trait_key=trait_1.trait_key, trait_value=new_trait_1_value
         ),
         generate_trait_data_item(trait_key=trait_3_key, trait_value=trait_3_value),
+        generate_trait_data_item(trait_key=trait_3_key, trait_value=trait_3_value),
```
Comment (matthewelwell) [inline api/tests/unit/environments/identities/test_unit_identities_models.py]: Hmm... it's odd to me that we're generating the same trait twice here?

File: api/tests/unit/environments/identities/test_unit_identities_views.py [hunk 5]
```diff
@@ -1126,6 +1143,59 @@ def test_post_identities__server_key_only_feature__server_key_auth__return_expec
...
+    }
+
+    # When
+    response = api_client.post(
+        url, data=json.dumps(data), content_type="application/json"
+    )
+
+    # Then
+    assert response.status_code == status.HTTP_200_OK
+    assert not Identity.objects.filter(identifier=identifier).count()
```
Comment (matthewelwell) [inline api/tests/unit/environments/identities/test_unit_identities_views.py]:
```suggestion
    assert not Identity.objects.filter(identifier=identifier).exists()
```

File: api/tests/unit/environments/identities/test_unit_identities_views.py [hunk 6]
```diff
@@ -1126,6 +1143,59 @@ def test_post_identities__server_key_only_feature__server_key_auth__return_expec
...
+    assert not Trait.objects.filter(trait_key=trait_key).count()
```
Comment (matthewelwell) [inline api/tests/unit/environments/identities/test_unit_identities_views.py]:
```suggestion
    assert not Trait.objects.filter(trait_key=trait_key).exists()
```

File: api/tests/unit/environments/identities/test_unit_identities_views.py [hunk 7]
```diff
@@ -1033,6 +1033,23 @@ def test_get_identities_with_hide_sensitive_data(
...
+    identifier = "transient"
+    api_client.credentials(HTTP_X_ENVIRONMENT_KEY=environment.api_key)
+    url = reverse("api-v1:sdk-identities") + f"?identifier={identifier}&transient=true"
+
+    # When
+    response = api_client.get(url)
+
+    # Then
+    assert response.status_code == status.HTTP_200_OK
+    assert not Identity.objects.filter(identifier=identifier).count()
```
Comment (matthewelwell) [inline api/tests/unit/environments/identities/test_unit_identities_views.py:1050]:
```suggestion
    assert not Identity.objects.filter(identifier=identifier).exists()
```

File: api/tests/unit/environments/identities/test_unit_identities_views.py [hunk 8]
```diff
@@ -1126,6 +1143,59 @@ def test_post_identities__server_key_only_feature__server_key_auth__return_expec
...
+
+    # When
+    response = api_client.post(
+        url, data=json.dumps(data), content_type="application/json"
+    )
+
+    # Then
+    assert response.status_code == status.HTTP_200_OK
+    assert Identity.objects.filter(identifier=identifier).count()
+    assert not Trait.objects.filter(trait_key=trait_key).count()
```
Comment (matthewelwell) [inline api/tests/unit/environments/identities/test_unit_identities_views.py]:
```suggestion
    assert Identity.objects.filter(identifier=identifier).exists()
    assert not Trait.objects.filter(trait_key=trait_key).exists()
```

File: api/tests/unit/environments/identities/traits/test_unit_traits_serializers.py [hunk 9]
```diff
@@ -48,7 +48,7 @@ def test_bulk_create_update_serializer_save_many(
     mocked_request = mocker.MagicMock(environment=identity.environment)
 
     # When
-    with django_assert_num_queries(6):
+    with django_assert_num_queries(5):
```
Comment (matthewelwell) [inline api/tests/unit/environments/identities/traits/test_unit_traits_serializers.py:51]: ❤️

**Review by matthewelwell — COMMENTED** [review]
I've added a few fairly minor comments, but one additional thing that (I think) is missing, is a test that verifies that transient traits are correctly used when evaluating segments / segment overrides. Maybe this is implicit from other tests but I certainly expected to see an explicit test for this case.

File: api/environments/identities/models.py (hunk 1 above)
Comment (khvn26) [reply api/environments/identities/models.py:267]: IMO no, it should not.

Comment (matthewelwell) [reply api/environments/identities/models.py:267]: I agree. But then I _think_ the code needs updating?

File: api/tests/unit/environments/identities/test_unit_identities_models.py (hunk 4 above)
Comment (khvn26) [reply api/tests/unit/environments/identities/test_unit_identities_models.py]: Fixed.

File: api/environments/identities/models.py (hunk 3 above)
Comment (khvn26) [reply api/environments/identities/models.py:317]: Yup!

File: api/environments/identities/models.py (hunk 1 above)
Comment (khvn26) [reply api/environments/identities/models.py:267]: This is done. Also added two integration tests as per the comment [here](https://github.com/Flagsmith/flagsmith/pull/4325#pullrequestreview-2175017859).

**Review by matthewelwell — APPROVED** [review]

**Pull request merged**